### PR TITLE
Моклев Вячеслав, 2539

### PR DESCRIPTION
**Pull request by slavam2605**
Убран всякий мусор, куча ненужных локальными переменных, развернут цикл update и все вынесено в отдельный поток (extends View и implements Runnable было очень странно О_о). Ну и отрисовка поменяна на drawBitmap(int [ ], ...)

### CODE REVIEW

File: app/src/main/java/ru/ifmo/md/lesson1/MainThread.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
+package ru.ifmo.md.lesson1;
+
+import android.graphics.Canvas;
+import android.util.Log;
+
+/**
+ * Created by Home on 14.09.2014.
+ */
+public class MainThread implements Runnable{
```
Comment (nwdeer) [inline app/src/main/java/ru/ifmo/md/lesson1/MainThread.java]: Bad naming. *Thread should extend a Thread class. If you are wrapping thread object, it should be ThreadWrapper or something like that.

File: app/src/main/java/ru/ifmo/md/lesson1/WhirlView.java [hunk 2]
```diff
@@ -1,72 +1,46 @@
...
-        running = false;
-        try {
-            thread.join();
-        } catch (InterruptedException ignore) {}
-    }
+        private final int width = 240;
+        private final int height = 320;
+        int [][] field = new int[width][height];
+        int [][] field2 = new int[width][height];
+        int [][] field3 = new int[width][height];
```
Comment (nwdeer) [inline app/src/main/java/ru/ifmo/md/lesson1/WhirlView.java]: is it used only for swapping? Pretty expensive solution keeping in mind that you might have only 16 mb of available memory. Can you come up with less memory consuming solution?

File: app/src/main/java/ru/ifmo/md/lesson1/WhirlView.java [hunk 3]
```diff
@@ -76,38 +50,91 @@ void initField() {
...
-    public void onDraw(Canvas canvas) {
-        for (int x=0; x<width; x++) {
-            for (int y=0; y<height; y++) {
-                Paint paint = new Paint();
-                paint.setColor(palette[field[x][y]]);
-                canvas.drawRect(x*scale, y*scale, (x+1)*scale, (y+1)*scale, paint);
+    public void render(Canvas canvas) {
+        for (int x = 0; x < width; x++) {
+            for (int y = 0; y < height; y++) {
+                colors[x + y * width] = palette[field[x][y]];
```
Comment (nwdeer) [inline app/src/main/java/ru/ifmo/md/lesson1/WhirlView.java:128]: Are you sure that updating array and than drawing it to canvas as bitmap is faster that drawing directly to canvas?

File: app/src/main/java/ru/ifmo/md/lesson1/WhirlView.java [hunk 4]
```diff
@@ -76,38 +50,91 @@ void initField() {
...
             }
         }
+        canvas.scale(scale_x, scale_y);
```
Comment (nwdeer) [inline app/src/main/java/ru/ifmo/md/lesson1/WhirlView.java:132]: If you are modifying canvas during onDraw method, you should use canvas.save/canvas.restore methods

File: app/src/main/java/ru/ifmo/md/lesson1/WhirlView.java (hunk 3 above)
Comment (slavam2605) [inline app/src/main/java/ru/ifmo/md/lesson1/WhirlView.java:128]: Yes, I'm totally sure, 17 ms bitmap->canvas vs ~100 ms to draw directly to canvas.

Comment (AVBelyy) [inline app/src/main/java/ru/ifmo/md/lesson1/WhirlView.java:128]: I agree with the previous speaker. My benchmarking confirms that as well. And I got 60 FPS!

**Comment by thevery** [comment]
ok = 8
~26 FPS

File: app/src/main/java/ru/ifmo/md/lesson1/MainThreadWrapper.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
+package ru.ifmo.md.lesson1;
+
+import android.graphics.Canvas;
+import android.util.Log;
+
+import java.util.Objects;
```
Comment (thevery) [inline app/src/main/java/ru/ifmo/md/lesson1/MainThreadWrapper.java:6]: unused import

File: app/src/main/java/ru/ifmo/md/lesson1/WhirlView.java [hunk 6]
```diff
@@ -1,72 +1,46 @@
...
+     */
+    class WhirlView extends SurfaceView {
 
-    public void pause() {
-        running = false;
-        try {
-            thread.join();
-        } catch (InterruptedException ignore) {}
-    }
+        private final int width = 240;
```
Comment (thevery) [inline app/src/main/java/ru/ifmo/md/lesson1/WhirlView.java:16]: `static final`

File: app/src/main/java/ru/ifmo/md/lesson1/WhirlView.java [hunk 7]
```diff
@@ -1,72 +1,46 @@
...
+        private final int height = 320;
+        int [][] field = new int[width][height];
+        int [][] field2 = new int[width][height];
+        int [][] field3 = null;
+        int [] colors;
+        float scale_x = 1;
+        float scale_y = 1;
```
Comment (thevery) [inline app/src/main/java/ru/ifmo/md/lesson1/WhirlView.java:23]: bad naming

File: app/src/main/java/ru/ifmo/md/lesson1/WhirlView.java [hunk 8]
```diff
@@ -76,38 +50,93 @@ void initField() {
...
-    @Override
-    public void onDraw(Canvas canvas) {
-        for (int x=0; x<width; x++) {
-            for (int y=0; y<height; y++) {
-                Paint paint = new Paint();
-                paint.setColor(palette[field[x][y]]);
-                canvas.drawRect(x*scale, y*scale, (x+1)*scale, (y+1)*scale, paint);
+    public void render(Canvas canvas) {
+        for (int x = 0; x < width; x++) {
+            for (int y = 0; y < height; y++) {
```
Comment (thevery) [inline app/src/main/java/ru/ifmo/md/lesson1/WhirlView.java:127]: you should probably "move" this loop inside `updateField`

**Pull request closed without merging**